pkg/operator/controller/externaldns/deployment.go

```diff
@@ -0,0 +1,261 @@
+	if !ok {
+		return nil, fmt.Errorf("%s is not a supported source type", externalDNS.Spec.Source.Type)
+	}
+	args = append(args, fmt.Sprintf("--source=%s", source))
+
+	if externalDNS.Spec.Source.Namespace != nil && len(*externalDNS.Spec.Source.Namespace) > 0 {
+		args = append(args, fmt.Sprintf("--namespace=%s", *externalDNS.Spec.Source.Namespace))
+	}
+
+	if externalDNS.Spec.Source.Service != nil && len(externalDNS.Spec.Source.Service.ServiceType) > 0 {
+		serviceTypeFilter, publishInternal := "", false
+		for _, serviceType := range externalDNS.Spec.Source.Service.ServiceType {
+			serviceTypeFilter += string(serviceType) + ","
+			if serviceType == corev1.ServiceTypeClusterIP {
+				publishInternal = true
+			}
+		}
+
+		// avoid having a trailing comma
+		args = append(args, fmt.Sprintf("--service-type-filter=%s", serviceTypeFilter[0:len(serviceTypeFilter)-1]))
+
+		if publishInternal {
+			args = append(args, "--publish-internal-services")
+		}
+	}
+
+	if externalDNS.Spec.Source.HostnameAnnotationPolicy == operatorv1alpha1.HostnameAnnotationPolicyIgnore {
+		args = append(args, "--ignore-hostname-annotation")
+	}
+
+	//TODO: Add logic for the CRD source.
+
+	return &corev1.Container{
+		Name:                     name,
+		Image:                    image,
+		ImagePullPolicy:          corev1.PullIfNotPresent,
+		TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
+		Args:                     args,
+	}, nil
+}
+
+// createExternalDNSDeployment creates the given deployment using the reconciler's client.
+func (r *reconciler) createExternalDNSDeployment(ctx context.Context, depl *appsv1.Deployment) error {
+	if err := r.client.Create(ctx, depl); err != nil {
+		return fmt.Errorf("failed to create externalDNS deployment %s/%s: %w", depl.Namespace, depl.Name, err)
+	}
+	r.log.Info("created externalDNS deployment", "namespace", depl.Namespace, "name", depl.Name)
+	return nil
+}
+
+// updateExternalDNSDeployment updates the in-cluster externalDNS deployment.
+// Returns a boolean if an update was made, and an error when relevant.
+func (r *reconciler) updateExternalDNSDeployment(ctx context.Context, current, desired *appsv1.Deployment) (bool, error) {
+	// don't always update (or do simple DeepEqual with) the operand's deployment
+	// as this may result into a "fight" between API/admission webhooks and this controller
+	// example:
+	//  - this contorller creates a deployment with the desired fields A, B, C
+	//  - API adds some default fields D, E, F (i.e. metadata, imagePullPullPolicy, dnsPolicy)
+	//  - mutating webhooks add default values to fields E, F
+	//  - this controller gets into reconcile loop and starts all from step1
+	//  - checking that fields A, B, C are the same as desired would save us the before mentioned round trips:
+	changed, updated := externalDNSDeploymentChanged(current, desired)
+	if !changed {
+		return false, nil
+	}
+
+	if err := r.client.Update(ctx, updated); err != nil {
```

Comment:
Shouldn't there be some kind of validation on the changes prior to making the update?

Reply:
Steve didn't do any validation and for the moment I cannot come up with any either. Also, I think that if something will come we'd better to put it into an admission webhook (validating or mutating for defaults).

Reply:
We usually do some validation around `Reconcile`, but it could be picked up in a later task.

Reply:
Yeah, you are right, these conditions need to be handled and set in the status. I think this can be attacked in the followups.

Reply:
Added the handling of the status to TOD in README.md in the [next PR](https://github.com/openshift/external-dns-operator/pull/20).